### ISSUE TITLE
Mini-v2 can not be produced in FastSim UL scenario

### ISSUE THREAD

**Issue by jordan-martins** [cms-sw/cmssw#34774]
Hi,

**[Description]:** we are closing the setup for RunII FastSim campaigns under UL scenarios. In our tests, Mini-v1 works just fine. However, Mini-v2 doesn't. 

**[Error]:**
`----- Begin Fatal Exception 04-Aug-2021 13:20:35 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'MINIAODSIMoutput_step'
   [2] Prefetching for module PoolOutputModule/'MINIAODSIMoutput'
   [3] Prefetching for module PATElectronSlimmer/'slimmedLowPtElectrons'
   [4] Prefetching for module PATElectronSelector/'selectedPatLowPtElectrons'
   [5] Prefetching for module PATElectronProducer/'patLowPtElectrons'
   [6] Calling method for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
Exception Message:
RefCore: A request to resolve a reference to a product of type 'std::vector<reco::Track>' with ProductID '1:982'
can not be satisfied because the product cannot be found.
Probably the branch containing the product is not stored in the input file.`

**[How to reproduce]:**
- Mini-v1 (working):
`wget https://cms-pdmv.cern.ch/mcm/public/restapi/requests/get_test/PPD-RunIISpring21UL16FSwmLHEGSPremix-00001`
this will produce AODSIM, since FastSim. For more quick test, edit the EVENTS variable of the script to have less events
`wget https://cms-pdmv.cern.ch/mcm/public/restapi/requests/get_test/PPD-RunIISummer20UL16MiniAOD-00009`
 
- Mini-v2 (issue):
`wget https://cms-pdmv.cern.ch/mcm/requests?prepid=PPD-RunIISummer20UL16MiniAODv2-00008&page=0&shown=127`
this will use the same AODSIM from the above (mini-v1) recipe

Could you kindly advise on how should we deal with mini-v2 scheme if possible at all considering that this is FastSim, please?

Thanks,
Jordan for PdmV

**Comment by slava77** [comment]
assign reconstruction

**Comment by slava77** [comment]
@bainbrid @afiqaize @SohamBhattacharya 
please check and clarify if you are available to take a look at this in the next days.
Thank you.

**Comment by bainbrid** [comment]
Hi @slava77 I’m away until the 16th Aug. 

My recollection was that low pT electrons are not supported for FastSim workflows, but the links below suggest otherwise …?

 https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronSeeds_cfi.py#L45

https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py#L62

**Comment by SohamBhattacharya** [comment]
Hi, @slava77 , I'm following up with this within the group and will get back soon.

**Comment by SohamBhattacharya** [comment]
Hi @jordan-martins ,
while @bainbrid , who is the expert on this, is away, I got some feedback from the group.
Would it be easy for you to check if setting `embedTrack = True` in [1] solves the problem?

[1] https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py#L62

**Comment by jordan-martins** [comment]
Hi @SohamBhattacharya , sure! Would you kindly provide how should I do this from the cmsdrive itself, please?
Thanks

**Comment by bainbrid** [comment]
Hi all,

I've tried the recipes provided [above](https://github.com/cms-sw/cmssw/issues/34774#issue-960576497) by @jordan-martins. 


These are the relevant modifiers/eras used for each test:

- PPD-RunIISpring21UL16FSwmLHEGSPremix-00001: `--era Run2_2016 --fast --procModifiers premix_stage2`

- PPD-RunIISummer20UL16MiniAOD-00009: `--era Run2_2016 --fast`

- PPD-RunIISummer20UL16MiniAODv2-00008: `--era Run2_2016 --fast --procModifiers run2_miniAOD_UL`

As far as I can tell, the problem is that here we are running a *re*-miniAOD of the FastSim workflow. I'm puzzled as to why this issue did not show up in any of the runTheMatrix tests. Perhaps we do not (yet) test this configuration?  

The `--procModifiers run2_miniAOD_UL` argument schedules the rerunning of the low pT electron ID module as part of the PAT process, see [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_25/PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py#L79). As a consequence, the offending line is [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_25/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L52). (I've tested the v2 without the `--procModifiers run2_miniAOD_UL` argument and it runs fine. I also did edmConfigDump and manually removed the low pT ID module from the task before running - again it runs fine.)

I'm quite sure we didn't ever plan to support low pT electrons for FastSim workflows, which is presumably why we see this problem. I will try to remember the details. But the problem is related to the information stored in the AOD collections, *not* in the PAT process itself.

And I'm thinking of how best to deal with this ... Opinions welcome. (@slava77?)

**Comment by jordan-martins** [comment]
Hi @bainbrid, thanks for this follow-up! 

From what I can tell, the relval wfs used Mini-v1 configuration and not Mini-v2. For example: https://cms-pdmv.cern.ch/relval/api/relvals/get_cmsdriver/CMSSW_10_6_22__fastsim_PU_2016premix-TTbar_13-00001 

Please, let us know how you would like to deal with this. From the PdmV side, we would like to have the Mini-v2 working solution (and not use Mini-v1), even if this means not have the low-pT electrons collection as I understand from above that it was the plan since the beginning.

Thanks,
Jordan

**Comment by bainbrid** [comment]
Hi @jordan-martins, @slava77, all,

some further investigations ...

The reason for the exception above is that for FastSim, we switch to a different track collection (from `generalTracks` to `generalTracksBeforeMixing`) to seed the low pT electron reconstruction (here's the [commit](https://github.com/cms-sw/cmssw/commit/1b3dd5721b7f74ea4b162b37a04a5ecc5ad68c00#diff-ee53734cba9db732577d612624ee1f80981d2fdcd3a4d64880103f6a8636ebc9)) and the `generalTracksBeforeMixing` collection _is not stored in the AOD event content_. Using the `run2_miniAOD_UL` modifier means that the ID module is rerun as part of the PAT sequence and it tries to deference a `TrackRef` to the missing `generalTracksBeforeMixing` collection. 

The reason we only see it now is because the ID module is rerun for the `run2_miniAOD_UL` modifier, and the use of this modifier in conjunction with fastsim wfs is (presumably?) not tested as part of runTheMatrix. 

I'm in touch with the author of the above change (who may be away currently) to understand better, and to find the best solution for future releases.

With regards to this particular production, perhaps the simplest change is to switch off the low pT electron part of the PAT sequence for the FastSim wfs? i.e. I can modify [this](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py#L126) logic to enable low pT electrons if, and only if, the wf is not FastSim? 

However, this would mean cutting a new release...? 

Are there other options, anyone?

**Comment by slava77** [comment]
@bainbrid 
is it possible to use the generalTracks in this context (separate PAT step running from AOD)?

**Comment by bainbrid** [comment]
hi @slava77, please can you rephrase - not sure I fully understand your question ...?

**Comment by bainbrid** [comment]
@jordan-martins how urgent is this issue?

**Comment by jordan-martins** [comment]
Hi @bainbrid , this is currently holding the startup of the FastSim UL campaigns. Not burning critical, but critical. 
Best, Jordan

**Comment by bainbrid** [comment]
Hi @jordan-martins, @Sam-Harper  @slava77, all,
 
> With regards to this particular production, perhaps the simplest change is to switch off the low pT electron part of the PAT sequence for the FastSim wfs? i.e. I can modify [this](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py#L126) logic to enable low pT electrons if, and only if, the wf is not FastSim?
> 

I’m currently away but I could probably implement the above one evening, as this seems to be the path of least resistance, but it would mean removing low pT ele from PAT for the FS wfs.

**Comment by slava77** [comment]
as it was discussed in the ORP earlier today, one of the main customers of fastsim are SUSY analyses and some would need the low-pt electrons supported.
So, a plain removal from the PAT sequence is not desirable.

@cms-sw/xpog-l2

**Comment by bainbrid** [comment]
> as it was discussed in the ORP earlier today, one of the main customers of fastsim are SUSY analyses and some would need the low-pt electrons supported.
> So, a plain removal from the PAT sequence is not desirable.
> 
> @cms-sw/xpog-l2

Ok then I’ll need to speak with @Sam-Harper on the details. Any significant devs may have to wait until next week.

**Comment by bainbrid** [comment]
Hi all, @slava77 @jordan-martins, 

@jordan-martins a quick (but perhaps important) question. I see your test uses CMSSW_10_6_25 for v2. Can I ask why CMSSW_10_6_26 is not used? Which release is being used for the v2 production in general? 

And then a quick update ...

- The solution for master (12_1_X) is trivial and I will provide this shortly. 
- I agree the solution provided above does not work *for 10_6_X* as it modifies the RECO chain. 
- I cannot simply reschedule the `lowPtGsfElectronCores` and `lowPtGsfElectrons` modules with an updated configuration (i.e. embed a track from `generalTracks` rather than `generalTracksBeforeMixing`) as there are input collections missing in AOD.
- I have written a trivial EDProducer (to be scheduled only for FastSim) that simply duplicates both the `lowPtGsfElectronCores` and `lowPtGsfElectrons` collections, *but* it embeds the correct track in the Core along [these](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/plugins/GsfElectronCoreBaseProducer.cc#L54-L55) lines, and it also embeds a Ref to the new Core in the Electron object. However, when I create a copy of the GsfElectron, this appears to trigger an exception due to a missing `TrackExtra` collection - I'm investigating, but it looks like this approach might also not work.
- My final fallback (but ugly hack) will be not use the `reco::GsfElectron::closestCtfTrackRef()` method ([here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L51)) in the IDProducer but instead use a track from `generalTracks` identified via the `lowPtGsfToTrackLinks` collection (which stores the provenance information for the low pT electrons). An example of this can already be found in the PAT, [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/plugins/LowPtGSFToPackedCandidateLinker.cc#L119), which allows the association lowPtGsfElectron -> lowPtGsfTrack -> reco::Track. This, strictly, is not necessarily a fully consistent sample of tracks as used by the ID training, but I suspect the performance penalty will be negligible (because, for a significant fraction of the electrons, the track will be the "correct" one).

**Comment by bainbrid** [comment]
Hi all, @slava77 @jordan-martins, 

And then a quick update ...

- The solution for master (12_1_X) is trivial and I will provide this shortly. 
- I agree the solution provided above does not work *for 10_6_X* as it modifies the RECO chain. 
- I cannot simply reschedule the `lowPtGsfElectronCores` and `lowPtGsfElectrons` modules with an updated configuration (i.e. embed a track from `generalTracks` rather than `generalTracksBeforeMixing`) as there are input collections missing in AOD.
- I have written a trivial EDProducer (to be scheduled only for FastSim) that simply duplicates both the `lowPtGsfElectronCores` and `lowPtGsfElectrons` collections, *but* it embeds the correct track in the Core along [these](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/plugins/GsfElectronCoreBaseProducer.cc#L54-L55) lines, and it also embeds a Ref to the new Core in the Electron object. However, when I create a copy of the GsfElectron, this appears to trigger an exception due to a missing `TrackExtra` collection - I'm investigating, but it looks like this approach might also not work.
- My final fallback (but ugly hack) will be not use the `reco::GsfElectron::closestCtfTrackRef()` method ([here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L51)) in the IDProducer but instead use a track from `generalTracks` identified via the `lowPtGsfToTrackLinks` collection (which stores the provenance information for the low pT electrons). An example of this can already be found in the PAT, [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/plugins/LowPtGSFToPackedCandidateLinker.cc#L119), which allows the association lowPtGsfElectron -> lowPtGsfTrack -> reco::Track. This, strictly, is not necessarily a fully consistent sample of tracks as used by the ID training, but I suspect the performance penalty will be negligible (because, for a significant fraction of the electrons, the track will be the "correct" one).

**Comment by jordan-martins** [comment]
> @jordan-martins a quick (but perhaps important) question. I see your test uses CMSSW_10_6_25 for v2. Can I ask why CMSSW_10_6_26 is not used? Which release is being used for the v2 production in general?

Hi @bainbrid, we currently use 10_6_25 for 20UL 16/16APV and 10_6_20 for 18 and 17.



> * My final fallback (but ugly hack) will be not use the `reco::GsfElectron::closestCtfTrackRef()` method ([here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L51)) in the IDProducer but instead use a track from `generalTracks` identified via the `lowPtGsfToTrackLinks` collection (which stores the provenance information for the low pT electrons). An example of this can already be found in the PAT, [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/plugins/LowPtGSFToPackedCandidateLinker.cc#L119), which allows the association lowPtGsfElectron -> lowPtGsfTrack -> reco::Track. This, strictly, is not necessarily a fully consistent sample of tracks as used by the ID training, but I suspect the performance penalty will be negligible (because, for a significant fraction of the electrons, the track will be the "correct" one).

I assume that we would need perhaps a technical validation to this (?). Or isn't it necessary?

Thanks a lot,
Jordan

**Comment by bainbrid** [comment]
Ok, a possible solution:

For master (12_1_X), I will simply revert part of [this](https://github.com/cms-sw/cmssw/commit/1b3dd5721b7f74ea4b162b37a04a5ecc5ad68c00#diff-ee53734cba9db732577d612624ee1f80981d2fdcd3a4d64880103f6a8636ebc9) commit, as follows:
https://github.com/cms-sw/cmssw/compare/master...bainbrid:LowPtElectrons_FastSimFix

For 10_6_X and the UL mini v2 FastSim campaign, I propose the following:

- I will remove the scheduling of the offending module `lowPtGsfElectronID` for the UL re-miniAOD FastSim workflows, which is a simple python change: https://github.com/cms-sw/cmssw/compare/CMSSW_10_6_X...bainbrid:LowPtElectrons_UL_FastSim_Fix_106X
- All other things remain the same, i.e. low pT electrons will still be available in the UL mini v2 FastSim samples. 
- The physics consequence of the above change is that the ID producer will not be rerun to apply the 2020Sept15 model, which is the default model for all other UL mini v2 workflows. Hence the low pT electrons will only be available with an obsolete (and not recommended) ID, "Autumn18". 

However, analysers will be able to obtain the latest ID model through two options:
1) If UL nano v9 is made available for FastSim samples, the latest "2020Nov28" ID model will be available as default in the NANOAOD. (I note here that this is encouraged: several other features are added as part of nano v9, such as energy regression, photon conversion vetoes, isolation, etc...)
2) The analysers can also re-run the latest ID model "on-the-fly" on top of the mini v2, as part of their analysis processing.

Please let me know if this is acceptable and I will make the PRs. 

@slava77 @mariadalfonso @gouskos

**Comment by bainbrid** [comment]
ok, after studies, the suggestion [above](https://github.com/cms-sw/cmssw/issues/34774#issuecomment-910735783) does not work, which leaves me only with the following option:

> * My final fallback (but ugly hack) will be not to use the `reco::GsfElectron::closestCtfTrackRef()` method ([here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L51)) in the IDProducer but instead use a track from `generalTracks` identified via the `lowPtGsfToTrackLinks` collection (which stores the provenance information for the low pT electrons). An example of this can already be found in the PAT, [here](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/PhysicsTools/PatAlgos/plugins/LowPtGSFToPackedCandidateLinker.cc#L119), which allows the association lowPtGsfElectron -> lowPtGsfTrack -> reco::Track. This, strictly, is not necessarily a fully consistent sample of tracks as used by the ID training, but I suspect the performance penalty will be negligible (because, for a significant fraction of the electrons, the track will be the "correct" one).

It's not as ugly as I thought and a compare for the proposed implementation can be found here:
https://github.com/cms-sw/cmssw/compare/CMSSW_10_6_X...bainbrid:LowPtElectrons_ULFastSimFix_106X

More details in today's RECO/AT meeting: https://indico.cern.ch/event/1071803/

**Comment by bainbrid** [comment]
Hi all, @slava77 @jordan-martins, I'd like to test my dev for master, using a recent IB for 12_1_X. Can anyone point me to scripts (as found here https://github.com/cms-sw/cmssw/issues/34774#issue-960576497) or runTheMatrix workflows that achieve the same steps for testing: i.e. GEN->AOD and then mini v2? Thanks!

**Comment by slava77** [comment]
> Hi all, @slava77 @jordan-martins, I'd like to test my dev for master, using a recent IB for 12_1_X. Can anyone point me to scripts (as found here [#34774 (comment)](https://github.com/cms-sw/cmssw/issues/34774#issue-960576497)) or runTheMatrix workflows that achieve the same steps for testing: i.e. GEN->AOD and then mini v2? Thanks!

I'd expect that in 12_X the test would use the AOD made in 10_6_X and run a regular re-miniAOD workflow, which would be the regular miniAOD config amended with `--procModifiers run2_miniAOD_UL`

**Comment by slava77** [comment]
for 2018, the matrix wf 1325.518 can be a reference; this needs to be spliced with the fastsim for the case at hand here.

**Comment by slava77** [comment]
@jordan-martins 
do we have feedback already from the fastsim UL pilots?
If the jobs are running fine, perhaps this issue can be closed.

**Comment by perrotta** [comment]
@jordan-martins @cms-sw/fastsim-l2 I think this issue can be closed now: please do so
